### PR TITLE
Expose OIDC and JWT events

### PR DESCRIPTION
**Pull request by laura-rodriguez**
Expose OIDC events for .NET 4.x and Core MVC projects.
Expose JWT events for .NET 4.x and Core WebApi projects.
Add tests.
Update readme.

### CODE REVIEW

File: Okta.AspNet.WebApi.IntegrationTest/MiddlewareShould.cs [hunk 1]
```diff
@@ -66,6 +66,20 @@ public async Task Returns401WhenAccessToProtectedRouteWithInvalidTokenAsync()
...
+        public async Task InvokeCustomEventsAsync()
+        {
+            var accessToken = "thisIsAnInvalidToken";
+            HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, ProtectedEndpoint);
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+
+            using (var client = new HttpClient(_server.Handler))
+            {
+                var response = await client.SendAsync(request);
+                Assert.True(response.Headers.Contains("myCustomHeader"));
```
Comment (bryanapellanes-okta) [inline Okta.AspNet.WebApi.IntegrationTest/MiddlewareShould.cs:79]: Where is "myCustomHeader" defined?

File: Okta.AspNet.WebApi.IntegrationTest/Startup.cs [hunk 2]
```diff
@@ -16,10 +20,18 @@ public class Startup
         public void Configuration(IAppBuilder app)
         {
             var oktaDomain = Environment.GetEnvironmentVariable("okta:OktaDomain");
-
+            var jwtProvider = new OAuthBearerAuthenticationProvider
+            {
+                OnApplyChallenge = context =>
+                {
+                    context.OwinContext.Response.Headers.Add("myCustomHeader", new [] {"myCustomValue"});
```
Comment (bryanapellanes-okta) [inline Okta.AspNet.WebApi.IntegrationTest/Startup.cs:27]: I think this is the answer to my previous question.

File: README.md [hunk 3]
```diff
@@ -5,11 +5,12 @@
...
-* [Need help?](#need-help)
-* [What you need](#what-you-need)
-* [Getting started](#getting-started)
-* [Contributing](#contributing)
+- [Okta ASP.NET middleware](#okta-aspnet-middleware)
+  - [Release status](#release-status)
+  - [Need help?](#need-help)
+  - [What you need](#what-you-need)
+  - [Getting Started](#getting-started)
+  - [Contributing](#contributing)
```
Comment (bryanapellanes-okta) [inline README.md:13]: Are these lines supposed to be indented?

File: docs/aspnetcore-mvc.md [hunk 4]
```diff
@@ -235,6 +235,7 @@ public IActionResult SignInWithIdp(string idp)
     return RedirectToAction("Index", "Home");
 }
 ```
+The Okta.AspNetCore library will include your identity provider id in the authorize URL and the user will prompted with the identity provider login. For more information, check out our guides to [add an external identity provider](https://developer.okta.com/docs/guides/add-an-external-idp/).
```
Comment (bryanapellanes-okta) [inline docs/aspnetcore-mvc.md:238]:
```suggestion
The Okta.AspNetCore library will includes your identity provider id in the authorize URL and the user is prompted with the identity provider login. For more information, check out our guides to [add an external identity provider](https://developer.okta.com/docs/guides/add-an-external-idp/).
```

File: docs/aspnet4x-mvc.md [hunk 5]
```diff
@@ -174,7 +174,11 @@ This example assumes you have a view called `Claim` whose model is of type `Syst
 
 ## Handling failures
 
-In the event a failure occurs, the Okta.AspNet library provides the `OnAuthenticationFailed` delegate defined on the `OktaMvcOptions` class. The following is an example of how to use `OnAuthenticationFailed` to handle authentication failures:
+library exposes [OpenIdConnectEvents](https://docs.microsoft.com/en-us/previous-versions/aspnet/mt180963(v=vs.113)) so you can hook into specific events during the authentication process. For more information see [`AuthenticationFailed`](https://docs.microsoft.com/en-us/previous-versions/aspnet/mt180967(v=vs.113)).
```
Comment (bryanapellanes-okta) [inline docs/aspnet4x-mvc.md:177]: Looks like the beginning of the sentence was removed.  Should it read `This library` or `The Okta.AspNet library`?

**Review by bryanapellanes-okta — APPROVED** [review]
The implementation looks good to me; some minor comments in the README.

File: Okta.AspNet.WebApi.IntegrationTest/MiddlewareShould.cs (hunk 1 above)
Comment (andriizhegurov-okta) [reply Okta.AspNet.WebApi.IntegrationTest/MiddlewareShould.cs:79]: I think it's here, line 27 https://github.com/okta/okta-aspnet/blob/2f053ab6133f901b0ff61d48d970f044b29d7781/Okta.AspNet.WebApi.IntegrationTest/Startup.cs#L20

File: docs/aspnetcore-mvc.md [hunk 6]
```diff
@@ -321,11 +325,8 @@ The `OktaMvcOptions` class configures the Okta middleware. You can see all the a
...
 | ClockSkew                 | No           | The clock skew allowed when validating tokens. The default value is 2 minutes. |
-| OnTokenValidated                 | No           | The event invoked after the security token has passed validation and a ClaimsIdentity has been generated. |
-| OnUserInformationReceived                 | No           | The event invoked when user information is retrieved from the UserInfoEndpoint. The `GetClaimsFromUserInfoEndpoint` value must be `true` when using this event. |
-| OnOktaApiFailure          | No           | The event invoked when a failure occurs within the Okta API. |
-| OnAuthenticationFailed    | No           | The event invoked if exceptions are thrown during request processing. |
 | Proxy                     | No           | An object describing proxy server configuration.  Properties are `Host`, `Port`, `Username` and `Password` |
+|OpenIdConnectEvents | No |  Specifies the [events](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.remoteauthenticationevents.onremotefailure) which the underlying OpenIdConnectHandler invokes to enable developer control over the authentication process.|
```
Comment (andriizhegurov-okta) [inline docs/aspnetcore-mvc.md:329]: This link to the MS docs looks not correct. Should be probably this https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.openidconnect.openidconnectevents

**Review by andriizhegurov-okta — APPROVED** [review]
👍

File: docs/aspnetcore-mvc.md (hunk 6 above)
Comment (laura-rodriguez) [reply docs/aspnetcore-mvc.md:329]: good catch 👍

File: docs/aspnet4x-mvc.md (hunk 5 above)
Comment (laura-rodriguez) [reply docs/aspnet4x-mvc.md:177]: yes, good catch 👍

File: README.md (hunk 3 above)
Comment (laura-rodriguez) [reply README.md:13]: I'll double checked.

**Pull request merged**